### PR TITLE
Custom index

### PR DESCRIPTION
**Pull request by jaimebuelta**
Adding command line option to add a custom index.

Some refactoring done to achieve it:

- Using argparse for command line arguments
- Using the definition of an index in HTML, as defined in PEP 503, instead of previous json (which is nicer, but not standard)
- Adding minimal test support with tox

### CODE REVIEW

File: wi [hunk 1]
```diff
@@ -54,17 +86,32 @@ async def download(url, dest):
...
+    for release in releases:
+        if 'http' in release['href']:
+            release['url'] = release['href']
+        else:
+            release['url'] = '/'.join([repo_url, release['href']])
+
+    # Filter by version, if required
+    if version:
+        releases = [r for r in releases if version in r['filename']]
+
```
Comment (chromano) [inline wi]: Perhaps:
```
if version:
    releases = filter(lambda release: release['filename'].rstrip('-', 1)[1].startswith(version),
        filter(lambda release: '-' in release, releases))
```

File: wi [hunk 2]
```diff
@@ -22,30 +22,62 @@ import asyncio
...
+    def handle_endtag(self, tag):
+        if self.current:
+            self.all_data.append(self.current)
+
+    def handle_data(self, data):
+        if self.current:
+            self.current['filename'] = data
+
+
 PACKAGES_REPO_URL = 'https://pypi.python.org/pypi'
```
Comment (chromano) [inline wi]: I think we can remove this constant now.

File: wi [hunk 3]
```diff
@@ -22,30 +22,62 @@ import asyncio
...
 
+
 def log(msg):
     sys.stderr.write(msg + '\n')
 
-async def metadata(package, version):
+
+async def metadata(repo_url, package, version):
     """Fetches metadata from the packages repository. Returns None
     if there's no package or package version available."""
```
[the remark this answers is not in the document]
Comment (jaimebuelta) [reply wi]: 👍

File: wi [hunk 4]
```diff
@@ -22,30 +22,62 @@ import asyncio
...
-        async with session.get('/'.join(url)) as response:
+        url = [repo_url, package]
+        async with session.request('GET', '/'.join(url)) as response:
             try:
-                return await response.json()
-            except ContentTypeError:
+                return await response.text()
+            except ContentTypeError as err:
+                log('Error retrieving {}'.format(url))
                 return
```
[the remark this answers is not in the document]
Comment (jaimebuelta) [reply wi]: Adding a test with a wrong index that returns 404

File: wi [hunk 5]
```diff
@@ -54,17 +86,32 @@ async def download(url, dest):
...
-    if not package:
-        return ('missing', name, version)
-    releases = package['releases'][package['info']['version']]
-    with tempfile.TemporaryDirectory() as tmpdir: 
+    package = await metadata(repo_url, name, version)
+    parser = PyPIParser()
+    parser.feed(package)
+    # if not package:
+    #     return ('missing', name, version)
+    # releases = package['releases'][package['info']['version']]
```
[the remark this answers is not in the document]
Comment (jaimebuelta) [reply wi]: Agreed. It's added now

File: wi (hunk 2 above)
Comment (jaimebuelta) [reply wi]: Removed

File: wi (hunk 1 above)
Comment (jaimebuelta) [reply wi]: I've changed it to a more explicit:

```
# The format of the wheel filename is, (PEP 427)
# {distribution}-{version}(-{build tag})?-{python tag}-
# {abi tag}-{platform tag}.whl
WHEEL_REGEX = r'(?P<distribution>[\w\d.]+)-(?P<version>[\d.]+)-(\S+).whl'
WHEEL_MATCH = re.compile(WHEEL_REGEX)

def check_version_in_filename(filename, version):
    """
    Checks that the wheel filename has the proper version
    """
    result = re.match(WHEEL_MATCH, filename)
    if not result:
        # Not a wheel file
        return False

    return result.groupdict()['version'] == version
...
    if version:
        releases = [r for r in releases
                    if check_version_in_filename(r['filename'], version)]
```

(I know about regexes, but I think in this case it makes an explicit parsing of the filename

Comment (jaimebuelta) [reply wi]: Another option is going for the more obvious
```
version = filename.split('-')[1]
```
But it looks more "magical" to me, but maybe I'm overthinking it...

**Pull request merged**